2009-05-08  Andrew Haley  <aph@redhat.com>

	* openjdk/jdk/src/share/native/sun/java2d/cmm/lcms/LCMS.c: Work
	around lcms breakage.  See
	https://bugs.openjdk.java.net/show_bug.cgi?id=100050

--- openjdk/jdk/src/share/native/sun/java2d/cmm/lcms/LCMS.c~	2009-04-24 08:34:31.000000000 +0100
+++ openjdk/jdk/src/share/native/sun/java2d/cmm/lcms/LCMS.c	2009-05-07 19:11:37.000000000 +0100
@@ -610,14 +610,19 @@
         return FALSE;
     }
 
-    if (!Icc->Grow(Icc, delta)) {
-        free(ptr);
-        if(isNew) {
-            Icc->TagCount--;
-        }
-        J2dRlsTraceLn(J2D_TRACE_ERROR,
-                      "_cmsModifyTagData: Icc->Grow() == FALSE");
-        return FALSE;
+    /* We change the size of Icc here only if we know it'll actually
+     * grow: if Icc is about to shrink we must wait until we've read
+     * the previous data.  */
+    if (delta > 0) {
+	if (!Icc->Grow(Icc, delta)) {
+	    free(ptr);
+	    if(isNew) {
+		Icc->TagCount--;
+	    }
+	    J2dRlsTraceLn(J2D_TRACE_ERROR,
+			  "_cmsModifyTagData: Icc->Grow() == FALSE");
+	    return FALSE;
+	}
     }
 
     /* Compute size of tag data before/after the modified tag */
@@ -680,6 +685,18 @@
     temp = TransportValue32(profileSize);
     Icc->Write(Icc, sizeof(icUInt32Number), &temp);
 
+    /* Shrink Icc, if needed.  */
+    if (delta < 0) {
+	if (!Icc->Grow(Icc, delta)) {
+	    free(ptr);
+	    if(isNew) {
+		Icc->TagCount--;
+	    }
+	    J2dRlsTraceLn(J2D_TRACE_ERROR,
+			  "_cmsModifyTagData: Icc->Grow() == FALSE");
+	    return FALSE;
+	}
+    }
 
     /* Adjust tag offsets: if the tag is new, we must account
        for the new tag table entry; otherwise, only those tags after
